providers/vultr/vultr.go

```diff
@@ -0,0 +1,342 @@
+	}
+
+	api.client = vultr.NewClient(api.token, nil)
+
+	// Validate token
+	_, err := api.client.GetAccountInfo()
+	if err != nil {
+		return nil, err
+	}
+
+	return api, nil
+}
+
+// GetDomainCorrections gets the corrections for a DomainConfig
+func (api *VultrApi) GetDomainCorrections(dc *models.DomainConfig) ([]*models.Correction, error) {
+	dc.Punycode()
+
+	ok, err := api.isDomainInAccount(dc.Name)
+	if err != nil {
+		return nil, err
+	}
+
+	if !ok {
+		return nil, fmt.Errorf("%s is not a domain in the Vultr account", dc.Name)
+	}
+
+	records, err := api.client.GetDNSRecords(dc.Name)
+	if err != nil {
+		return nil, err
+	}
+
+	curRecords := make([]*models.RecordConfig, len(records))
+	for i := range records {
+		r, err := toRecordConfig(dc, &records[i])
+		if err != nil {
+			return nil, err
+		}
+
+		curRecords[i] = r
+	}
+
+	differ := diff.New(dc)
+	_, create, delete, modify := differ.IncrementalDiff(curRecords)
+
+	corrections := []*models.Correction{}
+
+	for _, mod := range delete {
+		id := mod.Existing.Original.(*vultr.DNSRecord).RecordID
+		corrections = append(corrections, &models.Correction{
+			Msg: fmt.Sprintf("%s; Vultr RecordID: %v", mod.String(), id),
+			F: func() error {
+				return api.client.DeleteDNSRecord(dc.Name, id)
+			},
+		})
+	}
+
+	for _, mod := range create {
+		r := toVultrRecord(dc, mod.Desired)
+		corrections = append(corrections, &models.Correction{
+			Msg: mod.String(),
+			F: func() error {
+				return api.client.CreateDNSRecord(dc.Name, r.Name, r.Type, r.Data, r.Priority, r.TTL)
+			},
+		})
+	}
+
+	for _, mod := range modify {
+		id := mod.Existing.Original.(*vultr.DNSRecord).RecordID
+		r := toVultrRecord(dc, mod.Desired)
+		r.RecordID = id
+		corrections = append(corrections, &models.Correction{
+			Msg: fmt.Sprintf("%s; Vultr RecordID: %v", mod.String(), id),
+			F: func() error {
+				return api.client.UpdateDNSRecord(dc.Name, *r)
+			},
+		})
+	}
+
+	return corrections, nil
+}
+
+// GetNameservers gets the Vultr nameservers for a domain
+func (api *VultrApi) GetNameservers(domain string) ([]*models.Nameserver, error) {
+	return models.StringsToNameservers(defaultNS), nil
+}
+
+// EnsureDomainExists adds a domain to the Vutr DNS service if it does not exist
+func (api *VultrApi) EnsureDomainExists(domain string) error {
+	ok, err := api.isDomainInAccount(domain)
+	if err != nil {
+		return err
+	}
+
+	if !ok {
+		// Vultr requires an initial IP, use a dummy one
+		err := api.client.CreateDNSDomain(domain, "127.0.0.1")
+		if err != nil {
+			return err
+		}
+
+		ok, err := api.isDomainInAccount(domain)
+		if err != nil {
+			return err
+		}
+		if !ok {
+			return fmt.Errorf("Unexpected error adding domain %s to Vultr account", domain)
+		}
+	}
+
+	return nil
+}
+
+func (api *VultrApi) isDomainInAccount(domain string) (bool, error) {
+	domains, err := api.client.GetDNSDomains()
+	if err != nil {
+		return false, err
+	}
+
+	var vd *vultr.DNSDomain
+	for _, d := range domains {
```

Comment:
does their api do any kind of paging we need to worry about?

Reply:
Nope. Passes integration tests and I tried myself with 300 entries.